### PR TITLE
Feature MQTT offline persistence connection settings

### PR DESCRIPTION
**Pull request by bigmoby**
This PR is aimed at having better MQTT integration. The MQTT Paho client could be instantiated with buffering options on disconnect.

I've already signed the Lightbend Contributor License Agreement (CLA)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+## Changelog
```
Comment (juanjoDiaz) [inline CHANGELOG.md]: You don't have to change this file. It will be updated by the Alpakka team whenever they release.

File: mqtt/src/main/scala/akka/stream/alpakka/mqtt/settings.scala [hunk 2]
```diff
@@ -290,23 +301,26 @@ final class MqttConnectionSettings private (val broker: String,
...
-                   keepAliveInterval: FiniteDuration = keepAliveInterval,
-                   connectionTimeout: FiniteDuration = connectionTimeout,
-                   disconnectQuiesceTimeout: FiniteDuration = disconnectQuiesceTimeout,
-                   disconnectTimeout: FiniteDuration = disconnectTimeout,
-                   maxInFlight: Int = maxInFlight,
-                   mqttVersion: Int = mqttVersion,
-                   serverUris: immutable.Seq[String] = serverUris,
-                   sslHostnameVerifier: Option[javax.net.ssl.HostnameVerifier] = sslHostnameVerifier,
-                   sslProperties: Map[String, java.lang.String] = sslProperties): MqttConnectionSettings =
+  private def copy(
```
Comment (juanjoDiaz) [inline mqtt/src/main/scala/akka/stream/alpakka/mqtt/settings.scala]: You need to run `sbt test:compile` so all files as formatted correctly.

File: mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala [hunk 3]
```diff
@@ -98,14 +98,33 @@ class MqttSourceSpec
       connectionSettings.toString should include("ssl://localhost:1885")
       connectionSettings.toString should include("auth(username)=Some(mqttUser)")
     }
+
+    "allow MQTT buffering offline support persistence" in {
```
Comment (juanjoDiaz) [inline mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala]: This test doesn't really test the functionality. It simply test the `toString` method....

File: mqtt/src/test/scala/docs/scaladsl/MqttSinkSpec.scala [hunk 4]
```diff
@@ -41,6 +41,7 @@ class MqttSinkSpec
 
   val topic = "sink-spec/topic1"
   val topic2 = "sink-spec/topic2"
+  val topic3 = "sink-spec/topic3"
```
Comment (juanjoDiaz) [inline mqtt/src/test/scala/docs/scaladsl/MqttSinkSpec.scala]: Is this actually used anywhere?

File: mqtt/src/test/scala/docs/scaladsl/MqttSinkSpec.scala [hunk 5]
```diff
@@ -136,5 +137,6 @@ class MqttSinkSpec
       message.topic shouldBe msg.topic
       message.payload shouldBe msg.payload
     }
+
```
Comment (juanjoDiaz) [inline mqtt/src/test/scala/docs/scaladsl/MqttSinkSpec.scala]: Unnecessary empty line

File: mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala [hunk 6]
```diff
@@ -98,14 +98,33 @@ class MqttSourceSpec
...
+    }
   }
 
   "mqtt source" should {
     "consume unacknowledged messages from previous sessions using manualAck" in {
       import system.dispatcher
 
       val topic = "source-spec/manualacks"
-      val input = Vector("one", "two", "three", "four", "five")
+      val input: Seq[String] = Vector("one", "two", "three", "four", "five")
```
Comment (juanjoDiaz) [inline mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala]: Unnecessary change

File: mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala [hunk 7]
```diff
@@ -421,5 +440,6 @@ class MqttSourceSpec
       val elem = source2.runWith(Sink.head)
       elem.futureValue shouldBe MqttMessage(willTopic, ByteString("ohi"))
     }
+
```
Comment (juanjoDiaz) [inline mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala]: Unnecessary change

File: mqtt/src/main/scala/akka/stream/alpakka/mqtt/settings.scala [hunk 8]
```diff
@@ -262,6 +270,9 @@ final class MqttConnectionSettings private (val broker: String,
   def withSslProperties(value: java.util.Map[String, String]): MqttConnectionSettings =
     copy(sslProperties = value.asScala.toMap)
 
+  def withOfflinePersistenceSettings(value: MqttOfflinePersistenceSettings): MqttConnectionSettings =
```
Comment (juanjoDiaz) [inline mqtt/src/main/scala/akka/stream/alpakka/mqtt/settings.scala]: I'd change this to:
```
  def withOfflinePersistenceSettings(bufferSize: Int = 5000, deleteOldestMessage: Boolean = false, persistBuffer: 
Boolean = true): MqttConnectionSettings =
    copy(offlinePersistenceSettings = Option(MqttOfflinePersistenceSettings(bufferSize, deleteOldestMessage, persistBuffer)))
```
and make `MqttOfflinePersistenceSettings` private.

File: CHANGELOG.md (hunk 1 above)
Comment (bigmoby) [reply CHANGELOG.md]: Ops! I did it in order to document changes occurred, I'll remove immediately, sorry :-)

File: mqtt/src/main/scala/akka/stream/alpakka/mqtt/settings.scala (hunk 2 above)
Comment (bigmoby) [reply mqtt/src/main/scala/akka/stream/alpakka/mqtt/settings.scala]: It's really strange, I've invoked test task and automatically formatted by scalafmt.

File: mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala (hunk 7 above)
Comment (bigmoby) [reply mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala]: removed

File: mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala (hunk 3 above)
Comment (bigmoby) [reply mqtt/src/test/scala/docs/scaladsl/MqttSourceSpec.scala]: Yes, I'll implement an effective test

**Comment by 2m** [comment]
Thanks @bigmoby for the feature and @juanjoDiaz for jumping on it for a review!

**Comment by 2m** [comment]
If you want to try this out in your codebase, you case use a snapshot that has been published by Travis after this PR got merged. The version is `1.0-M1+13-ff5cc18b`. More information on snapshots here: https://developer.lightbend.com/docs/alpakka/current/other-docs/snapshots.html
